File: Sources/Atlas/UI/MatrixView.swift
import SwiftUI

/// The dependency matrix: rows call columns.
///
/// No crossing lines at any size, so clusters, layers and cycles are read
/// directly off the grid rather than traced through a drawing. A cycle is not
/// inferred from which side of the diagonal a mark falls on — the reciprocal
/// pair is marked in magenta outright, which is both more direct and correct
/// for a pair sitting in the same district.
struct MatrixView: View {
    @EnvironmentObject private var state: AppState
    @EnvironmentObject private var loc: Localization

    @State private var selected: Int?

    private let cell: CGFloat = 15
    private let gap: CGFloat = 9
    private let leftGutter: CGFloat = 168
    private let topGutter: CGFloat = 152
    private let barWidth: CGFloat = 62

    var body: some View {
        ScrollView([.horizontal, .vertical]) {
            VStack(alignment: .leading, spacing: 16) {
                header
                Canvas { context, size in
                    draw(context: context, size: size)
                }
                .frame(width: canvas.width, height: canvas.height)
                .contentShape(Rectangle())
                .onTapGesture { point in
                    if let hit = index(at: point) { choose(hit) }
                }
                footer
            }
            .padding(26)
        }
        .background(PaperBackground())
        .onChange(of: state.fileGraph.nodes.count) { _, _ in selected = nil }
        .onChange(of: state.graph?.rootPath) { _, _ in selected = nil }
    }

    // MARK: - Ordering

    /// Grouping and row order come from `MatrixOrder` in the engine, so the
    /// Windows client lays out the same grid. Only the wording and the ink are
    /// decided here.
    private var ordered: MatrixOrder { MatrixOrder(graph: state.fileGraph) }

    private func label(_ district: MatrixOrder.District) -> String {
        switch district.key {
        case "interface": return loc.t.districtInterface
        case "data":      return loc.t.districtData
        default:          return loc.t.districtLogic
        }
    }

    private func ink(_ district: MatrixOrder.District) -> Color {
        switch district.key {
        case "interface": return Theme.inkCyan
        case "data":      return Theme.inkMagenta
        default:          return Theme.textSecondary
        }
    }

    private func position(_ rank: Int, in order: MatrixOrder) -> CGFloat {
        var gaps = 0
        for district in order.districts.dropFirst() where rank >= district.start { gaps += 1 }
        return CGFloat(rank) * cell + CGFloat(gaps) * gap
    }

    private var canvas: CGSize {
        let order = ordered
        guard !order.indices.isEmpty else { return CGSize(width: 400, height: 260) }
        let span = position(order.indices.count - 1, in: order) + cell
        return CGSize(width: leftGutter + span + barWidth + 26, height: topGutter + span + 34)
    }

    // MARK: - Drawing

    private func draw(context: GraphicsContext, size: CGSize) {
        let graph = state.fileGraph
        let order = ordered
        guard !order.indices.isEmpty else { return }
        let count = order.indices.count
        let span = position(count - 1, in: order) + cell

        var calls = Set<Int64>()
        for edge in graph.edges {
            guard let r = order.rank[edge.from], let c = order.rank[edge.to] else { continue }
            calls.insert(Int64(r) << 32 | Int64(c))
        }
        func has(_ r: Int, _ c: Int) -> Bool { calls.contains(Int64(r) << 32 | Int64(c)) }

        let selRank = selected.flatMap { order.rank[$0] }

        // ---- Bands for the selection: the row it calls in cyan, the column
        // that calls it in magenta. The same pairing as everywhere else.
        if let rank = selRank {
            context.fill(Path(CGRect(x: leftGutter, y: topGutter + position(rank, in: order),
                                     width: span, height: cell)),
                         with: .color(Theme.inkCyan.opacity(0.11)))
            context.fill(Path(CGRect(x: leftGutter + position(rank, in: order), y: topGutter,
                                     width: cell, height: span)),
                         with: .color(Theme.inkMagenta.opacity(0.09)))
        }

        // ---- Cells ----
        for r in 0..<count {
            for c in 0..<count {
                let x = leftGutter + position(c, in: order)
                let y = topGutter + position(r, in: order)

                if r == c {
                    // The file itself: a small neutral square, well inside the
                    // cell so it never reads as data.
                    context.fill(Path(CGRect(x: x + 5, y: y + 5, width: 5, height: 5)),
                                 with: .color(Theme.border))
                    continue
                }
                guard has(r, c) else { continue }

                let reciprocal = has(c, r)
                let live = selRank == r || selRank == c
                let ink = reciprocal ? Theme.inkMagenta : Theme.inkCyan
                context.fill(Path(CGRect(x: x + 1, y: y + 1, width: 13, height: 13)),
                             with: .color(ink.opacity(live ? 1 : 0.78)))
            }
        }

        // ---- Fan bars: how many files each row pulls in ----
        let outDegree = order.indices.map { graph.outgoing[$0].count }
        let maxOut = max(outDegree.max() ?? 1, 1)
        for (rank, degree) in outDegree.enumerated() {
            let width = CGFloat(degree) / CGFloat(maxOut) * barWidth
            guard width > 0.5 else { continue }
            context.fill(Path(CGRect(x: leftGutter + span + 10,
                                     y: topGutter + position(rank, in: order) + 4,
                                     width: width, height: 7)),
                         with: .color(Theme.borderStrong))
        }

        // ---- Row labels ----
        for (rank, node) in order.indices.enumerated() {
            let selectedRow = selRank == rank
            let text = Text(graph.nodes[node].name)
                .font(.system(size: 10.5, design: .serif))
                .foregroundStyle(selectedRow ? Theme.textPrimary : Theme.textSecondary)
            context.draw(context.resolve(text),
                         at: CGPoint(x: leftGutter - 10,
                                     y: topGutter + position(rank, in: order) + cell / 2),
                         anchor: .trailing)
        }

        // ---- Column labels, turned ----
        for (rank, node) in order.indices.enumerated() {
            let selectedCol = selRank == rank
            var turned = context
            turned.translateBy(x: leftGutter + position(rank, in: order) + 11, y: topGutter - 6)
            turned.rotate(by: .degrees(-90))
            let text = Text(graph.nodes[node].name)
                .font(.system(size: 10.5, design: .serif))
                .foregroundStyle(selectedCol ? Theme.textPrimary : Theme.textSecondary)
            turned.draw(turned.resolve(text), at: .zero, anchor: .leading)
        }

        // ---- District marks, below the grid ----
        for district in order.districts {
            let text = Text(label(district).uppercased())
                .font(.system(size: 10, weight: .semibold, design: .serif))
                .foregroundStyle(ink(district))
            context.draw(context.resolve(text),
                         at: CGPoint(x: leftGutter + position(district.start, in: order),
                                     y: topGutter + span + 12),
                         anchor: .topLeading)
        }
    }

    // MARK: - Interaction

    private func index(at point: CGPoint) -> Int? {
        let order = ordered
        for (rank, node) in order.indices.enumerated() {
            let o = position(rank, in: order)
            if point.x < leftGutter, point.y >= topGutter + o, point.y < topGutter + o + cell {
                return node
            }
            if point.y < topGutter, point.x >= leftGutter + o, point.x < leftGutter + o + cell {
                return node
            }
            if point.y >= topGutter + o, point.y < topGutter + o + cell, point.x >= leftGutter {
                return node
            }
        }
        return nil
    }

    private func choose(_ node: Int) {
        selected = node
        if let first = state.fileGraph.nodes[node].symbols.first { state.select(first) }
    }

    // MARK: - Chrome

    private var header: some View {
        HStack(alignment: .bottom, spacing: 24) {
            VStack(alignment: .leading, spacing: 3) {
                Text(loc.t.matrixTitle)
                    .font(Theme.Font.heading)
                    .foregroundStyle(Theme.textPrimary)
                Text(loc.t.matrixHint)
                    .font(Theme.Font.micro.weight(.regular))
                    .foregroundStyle(Theme.textTertiary)
                    .fixedSize(horizontal: false, vertical: true)
                    .frame(maxWidth: 520, alignment: .leading)
            }
            Spacer(minLength: 0)
            HStack(spacing: 16) {
                key(Theme.inkCyan, loc.t.matrixCallsKey)
                key(Theme.inkMagenta, loc.t.matrixCycleKey)
                key(Theme.border, loc.t.matrixSelfKey)
            }
        }
    }

    private func key(_ ink: Color, _ text: String) -> some View {
        HStack(spacing: 6) {
            Rectangle().fill(ink).frame(width: 10, height: 10)
            Text(text).font(Theme.Font.micro.weight(.regular)).foregroundStyle(Theme.textSecondary)
        }
    }

    private var footer: some View {
        HStack(alignment: .top, spacing: 34) {
            VStack(alignment: .leading, spacing: 2) {
                Text(loc.t.selectedRow.uppercased())
                    .font(Theme.Font.label).tracking(1.0)
                    .foregroundStyle(Theme.textTertiary)
                if let sel = selected, sel < state.fileGraph.nodes.count {
                    Text(state.fileGraph.nodes[sel].name)
                        .font(Theme.Font.mono)
                        .foregroundStyle(Theme.textPrimary)
                    Text(loc.t.matrixCounts(out: state.fileGraph.outgoing[sel].count,
                                            in: state.fileGraph.incoming[sel].count))
                        .font(Theme.Font.micro.weight(.regular))
                        .foregroundStyle(Theme.textTertiary)
                } else {
                    Text(loc.t.clickAnyFile)
                        .font(Theme.Font.caption)
                        .foregroundStyle(Theme.textTertiary)
                }
            }
            .frame(width: 250, alignment: .leading)

            Text(loc.t.matrixFooter)
                .font(Theme.Font.caption)
                .foregroundStyle(Theme.textSecondary)
                .lineSpacing(2.5)
                .fixedSize(horizontal: false, vertical: true)
                .frame(maxWidth: 560, alignment: .leading)
        }
    }
}

File: Windows/Atlas.Windows/Views/MatrixView.cs
using System.Globalization;
using Avalonia;
using Avalonia.Controls;
using Avalonia.Input;
using Avalonia.Media;
using Atlas.Windows.Engine;

namespace Atlas.Windows.Views;

/// <summary>
/// The dependency matrix: rows call columns.
///
/// No crossing lines at any size, so clusters, layers and cycles are read
/// directly off the grid rather than traced through a drawing. A cycle is not
/// inferred from which side of the diagonal a mark falls on — the reciprocal
/// pair is marked in magenta outright, which is both more direct and correct
/// for a pair sitting in the same district.
///
/// Ported from Sources/Atlas/UI/MatrixView.swift. The row order comes from
/// the engine's MatrixOrder, the same code the macOS build uses.
/// </summary>
public sealed class MatrixView : Control
{
    private const double Cell = 15;
    private const double Gap = 9;
    private const double LeftGutter = 168;
    private const double TopGutter = 152;
    private const double BarWidth = 62;

    private readonly Strings _t;
    private readonly MapInfo _map;
    private readonly IReadOnlyList<int> _order;
    private readonly Dictionary<int, int> _rank;
    private readonly HashSet<long> _calls;
    private readonly int[] _outDegree;

    private int? _selected;

    public MatrixView(Report report, Strings t)
    {
        _t = t;
        _map = report.Map ?? new MapInfo();
        _order = _map.Matrix.Order;

        _rank = new Dictionary<int, int>(_order.Count);
        for (int i = 0; i < _order.Count; i++) _rank[_order[i]] = i;

        // Membership by packed rank pair: the grid asks "is there a mark here"
        // once per cell, and a project of sixty files asks it 3,600 times a frame.
        _calls = [];
        foreach (var edge in _map.Edges)
        {
            if (_rank.TryGetValue(edge.From, out int r) && _rank.TryGetValue(edge.To, out int c))
                _calls.Add(Key(r, c));
        }

        _outDegree = new int[_order.Count];
        foreach (var edge in _map.Edges)
        {
            if (_rank.TryGetValue(edge.From, out int r)) _outDegree[r]++;
        }

        Focusable = true;
        if (Screenshot.Select is { } node && _rank.ContainsKey(node)) _selected = node;
    }

    private static long Key(int row, int column) => ((long)row << 32) | (uint)column;
    private bool Has(int row, int column) => _calls.Contains(Key(row, column));

    private static double Position(int rank, IReadOnlyList<MatrixDistrict> districts)
    {
        int gaps = 0;
        for (int i = 1; i < districts.Count; i++)
        {
            if (rank >= districts[i].Start) gaps++;
        }
        return rank * Cell + gaps * Gap;
    }

    private double Span =>
        _order.Count == 0 ? 0 : Position(_order.Count - 1, _map.Matrix.Districts) + Cell;

    protected override Size MeasureOverride(Size availableSize) =>
        _order.Count == 0
            ? new Size(400, 260)
            : new Size(LeftGutter + Span + BarWidth + 26, TopGutter + Span + 34);

    // MARK: - Interaction

    protected override void OnPointerPressed(PointerPressedEventArgs e)
    {
        base.OnPointerPressed(e);
        var point = e.GetPosition(this);
        var districts = _map.Matrix.Districts;

        for (int rank = 0; rank < _order.Count; rank++)
        {
            double o = Position(rank, districts);
            bool onRow = point.Y >= TopGutter + o && point.Y < TopGutter + o + Cell;
            bool onColumn = point.X >= LeftGutter + o && point.X < LeftGutter + o + Cell;
            if ((onRow && point.X < LeftGutter)
                || (onColumn && point.Y < TopGutter)
                || (onRow && point.X >= LeftGutter))
            {
                int node = _order[rank];
                _selected = _selected == node ? null : node;
                InvalidateVisual();
                return;
            }
        }
        _selected = null;
        InvalidateVisual();
    }

    // MARK: - Drawing

    public override void Render(DrawingContext context)
    {
        if (_order.Count == 0) return;

        var districts = _map.Matrix.Districts;
        int count = _order.Count;
        double span = Span;
        int? selectedRank = _selected is { } picked && _rank.TryGetValue(picked, out int r)
            ? r : null;

        // Bands for the selection: the row it calls in cyan, the column that
        // calls it in magenta. The same pairing as everywhere else.
        if (selectedRank is { } rank)
        {
            double at = Position(rank, districts);
            context.FillRectangle(Broadsheet.Brush(Broadsheet.Fade(Broadsheet.InkCyan, 0.11)),
                new Rect(LeftGutter, TopGutter + at, span, Cell));
            context.FillRectangle(Broadsheet.Brush(Broadsheet.Fade(Broadsheet.InkMagenta, 0.09)),
                new Rect(LeftGutter + at, TopGutter, Cell, span));
        }

        // ---- Cells ----
        var diagonalBrush = Broadsheet.Brush(Broadsheet.Border);
        for (int row = 0; row < count; row++)
        {
            double y = TopGutter + Position(row, districts);
            for (int column = 0; column < count; column++)
            {
                double x = LeftGutter + Position(column, districts);

                if (row == column)
                {
                    // The file itself: a small neutral square, well inside the
                    // cell so it never reads as data.
                    context.FillRectangle(diagonalBrush, new Rect(x + 5, y + 5, 5, 5));
                    continue;
                }
                if (!Has(row, column)) continue;

                bool reciprocal = Has(column, row);
                bool live = selectedRank == row || selectedRank == column;
                var ink = reciprocal ? Broadsheet.InkMagenta : Broadsheet.InkCyan;
                context.FillRectangle(
                    Broadsheet.Brush(Broadsheet.Fade(ink, live ? 1 : 0.78)),
                    new Rect(x + 1, y + 1, 13, 13));
            }
        }

        // ---- Fan bars: how many files each row pulls in ----
        int maxOut = Math.Max(_outDegree.Length == 0 ? 1 : _outDegree.Max(), 1);
        var barBrush = Broadsheet.Brush(Broadsheet.BorderStrong);
        for (int i = 0; i < _outDegree.Length; i++)
        {
            double width = (double)_outDegree[i] / maxOut * BarWidth;
            if (width <= 0.5) continue;
            context.FillRectangle(barBrush,
                new Rect(LeftGutter + span + 10, TopGutter + Position(i, districts) + 4,
                         width, 7));
        }

        // ---- Row labels ----
        for (int i = 0; i < count; i++)
        {
            var label = Label(_order[i], selectedRank == i);
            context.DrawText(label,
                new Point(LeftGutter - 10 - label.Width,
                          TopGutter + Position(i, districts) + Cell / 2 - label.Height / 2));
        }

        // ---- Column labels, turned ----
        for (int i = 0; i < count; i++)
        {
            var label = Label(_order[i], selectedRank == i);
            var at = new Point(LeftGutter + Position(i, districts) + 11, TopGutter - 6);
            using (context.PushTransform(
                       Matrix.CreateRotation(-Math.PI / 2) * Matrix.CreateTranslation(at.X, at.Y)))
            {
                context.DrawText(label, new Point(0, -label.Height / 2));
            }
        }

        // ---- District marks, below the grid ----
        foreach (var district in districts)
        {
            var text = Text(DistrictLabel(district.Key).ToUpperInvariant(), 10,
                            DistrictInk(district.Key), FontWeight.SemiBold);
            context.DrawText(text,
                new Point(LeftGutter + Position(district.Start, districts), TopGutter + span + 12));
        }
    }

    private FormattedText Label(int node, bool selected) =>
        Text(node >= 0 && node < _map.Nodes.Count ? _map.Nodes[node].Name : "",
             10.5, selected ? Broadsheet.TextPrimary : Broadsheet.TextSecondary);

    private string DistrictLabel(string key) => key switch
    {
        "interface" => _t["districtInterface"],
        "data"      => _t["districtData"],
        _           => _t["districtLogic"],
    };

    private static Color DistrictInk(string key) => key switch
    {
        "interface" => Broadsheet.InkCyan,
        "data"      => Broadsheet.InkMagenta,
        _           => Broadsheet.TextSecondary,
    };

    private static FormattedText Text(string value, double size, Color ink,
                                      FontWeight weight = FontWeight.Normal) =>
        new(value, CultureInfo.CurrentCulture, FlowDirection.LeftToRight,
            new Typeface(Broadsheet.Fonts.Serif, FontStyle.Normal, weight),
            size, Broadsheet.Brush(ink));
}
